### PR TITLE
net: lwm2m: Update security object to 1.1

### PR DESCRIPTION
**Pull request by jarlamsa**
Update LwM2M security object to version 1.1
Added Kconfig option to switch between object versions 1.0 and 1.1

Signed-off-by: Jarno Lamsa <jarno.lamsa@nordicsemi.no>

### CODE REVIEW

**Comment by jarlamsa** [comment]
@VeijoPesonen @jheiskan81 @SeppoTakalo  please review

File: subsys/net/lib/lwm2m/lwm2m_obj_security.c [hunk 1]
```diff
@@ -34,8 +45,13 @@ LOG_MODULE_REGISTER(LOG_MODULE_NAME);
 #define SECURITY_SHORT_SERVER_ID		10
 #define SECURITY_CLIENT_HOLD_OFF_TIME_ID	11
 #define SECURITY_BS_SERVER_ACCOUNT_TIMEOUT_ID	12
-
-#define SECURITY_MAX_ID				13
+#if defined(CONFIG_LWM2M_SECURITY_OBJECT_VERSION_1_1)
+#define SECURITY_MATCHING_TYPE_ID		13
+#define SECURITY_SNI_ID				14
+#define SECURITY_CERTIFICATE_USAGE_ID		15
```
Comment (VeijoPesonen) [inline subsys/net/lib/lwm2m/lwm2m_obj_security.c:51]: Extra whitespace.

File: subsys/net/lib/lwm2m/lwm2m_obj_security.c [hunk 2]
```diff
@@ -72,7 +96,14 @@ static struct lwm2m_engine_obj_field fields[] = {
 	OBJ_FIELD_DATA(SECURITY_LWM2M_SERVER_SMS_NUM_ID, W_OPT, STRING),
 	OBJ_FIELD_DATA(SECURITY_SHORT_SERVER_ID, W_OPT, U16),
 	OBJ_FIELD_DATA(SECURITY_CLIENT_HOLD_OFF_TIME_ID, W_OPT, U32),
-	OBJ_FIELD_DATA(SECURITY_BS_SERVER_ACCOUNT_TIMEOUT_ID, W_OPT, U32)
+	OBJ_FIELD_DATA(SECURITY_BS_SERVER_ACCOUNT_TIMEOUT_ID, W_OPT, U32),
```
Comment (VeijoPesonen) [inline subsys/net/lib/lwm2m/lwm2m_obj_security.c:99]: I don't know does those really matter but the following fields are defined as signed integers - S32s - by the specification

image

**Review by VeijoPesonen — CHANGES_REQUESTED** [review]
The changes look good but it seems that there are potential minor bugs adjacent to your modifications.

File: subsys/net/lib/lwm2m/lwm2m_obj_security.c (hunk 2 above)
Comment (jarlamsa) [reply subsys/net/lib/lwm2m/lwm2m_obj_security.c:99]: Good catch. First of those 2 are already in the version 1.0, but have probably seen no use as they aren't initialized at all in the object instance creation.

**Comment by jarlamsa** [comment]
Thank you @VeijoPesonen for your review comments. Those have been now addressed, please re-review.

**Review by VeijoPesonen — APPROVED** [review]

File: subsys/net/lib/lwm2m/lwm2m_obj_security.c [hunk 3]
```diff
@@ -46,8 +62,16 @@ LOG_MODULE_REGISTER(LOG_MODULE_NAME);
 /*
  * Calculate resource instances as follows:
  * start with SECURITY_MAX_ID
+ *
+ * If using object version 1.1
+ * subtract MULTI resources because their counts include 0 resource (1)
```
Comment (rlubos) [inline subsys/net/lib/lwm2m/lwm2m_obj_security.c:67]: I don't see this being applied in the actual code (line 71).

File: subsys/net/lib/lwm2m/lwm2m_obj_security.c [hunk 4]
```diff
@@ -138,6 +169,13 @@ static struct lwm2m_engine_obj_inst *security_create(uint16_t obj_inst_id)
 	INIT_OBJ_RES_DATA(SECURITY_SHORT_SERVER_ID, res[index], i,
 			  res_inst[index], j,
 			  &short_server_id[index], sizeof(*short_server_id));
+#if defined(CONFIG_LWM2M_SECURITY_OBJECT_VERSION_1_1)
+	INIT_OBJ_RES_OPTDATA(SECURITY_MATCHING_TYPE_ID, res[index], i, res_inst[index], j);
+	INIT_OBJ_RES_OPTDATA(SECURITY_SNI_ID, res[index], i, res_inst[index], j);
+	INIT_OBJ_RES_OPTDATA(SECURITY_CERTIFICATE_USAGE_ID, res[index], i, res_inst[index], j);
+	INIT_OBJ_RES_OPTDATA(SECURITY_DTLS_TLS_CIPHERSUITE_ID, res[index], i, res_inst[index], j);
```
Comment (rlubos) [inline subsys/net/lib/lwm2m/lwm2m_obj_security.c:176]: I think this should use `INIT_OBJ_RES_MULTI_OPTDATA` since it's a multi-instance resource.

**Comment by jarlamsa** [comment]
Thank you for your review comments @rlubos I think I have addressed those, could you please re-review.

**Review by rlubos — APPROVED** [review]
LGTM

**Review by jukkar — APPROVED** [review]

**Pull request merged**